Fixes #206 and a safety issue with requests.

imsenthur opened the pull request:
## Description

1) Moved temperature checks to actuators' update method from osl
2) Moved set and get methods to actuators from joints

## Related Issue

#206 

## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [x] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/imsenthur/opensourceleg/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/imsenthur/opensourceleg/blob/master/CONTRIBUTING.md) guide.
- [x] I've updated the code style using `make codestyle`.
- [x] I've written tests for all new methods and classes that I created.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

tkevinbest commented on opensourceleg/hardware/actuators.py:
```diff
@@ -497,6 +501,20 @@ def update(self) -> None:
+                self.stop()
+                exit()
+
+            if self.winding_temperature > self.max_winding_temperature:
```
Am I missing where `self.winding_temperature` is unpacked from the thermal model?

imsenthur replied:
The `winding_temperature` is unpacked here: https://github.com/neurobionics/opensourceleg/blob/b74c342cec77111ef06715b8355984706f1dd256/opensourceleg/hardware/actuators.py#L877-L885

tkevinbest commented on opensourceleg/hardware/actuators.py:
```diff
@@ -687,6 +705,24 @@ def set_motor_position(self, position: float) -> None:
             ),
         )
 
+    def set_max_case_temperature(self, temperature: float) -> None:
```
Is there a reason to do this versus the property getter/setter syntax? 
```
@max_case_temperature.setter
def max_case_temperature(self, temperature: float) -> None:
"""
        Set the maximum temperature of the motor.
        Args:
            temperature (float): temperature in degrees Celsius
        """
        self._max_case_temperature = temperature
```

imsenthur replied:
Most elements of the library follow the `set_variable(value)` routine instead of the `variable = value` or setter/getter way and IMO calling a method to set value to a critical variable is a better way to handle this instead of letting the user use assignment statements.

tkevinbest replied:
Got it, sounds good!

tkevinbest commented:
I've traced the logic through and everything looks good. I'm gonna give it a test on hardware soon where I make sure that it trips and everything shuts down correctly and then approve it.

tkevinbest commented:
Overall the functionality works. However, the `self.stop` command closes the actuator connection and then when the exit routine of the `with osl:` block exits again, you get an error: 
```
[2024-06-26 13:57:04,209] WARNING: [KNEE] Winding thermal limit 30 reached. Stopping motor.
Traceback (most recent call last):
  File "/home/pi/test_thermal_cutoffs.py", line 24, in <module>
    osl.update()
  File "/home/pi/opensourceleg/opensourceleg/osl.py", line 331, in update
    self._knee.update()
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 516, in update
    exit()
  File "/usr/lib/python3.9/_sitebuiltins.py", line 26, in __call__
    raise SystemExit(code)
SystemExit: None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/test_thermal_cutoffs.py", line 26, in <module>
    print("Case {:2f}, Winding {:2f}".format(osl.knee.case_temperature, osl.knee.winding_temperature))
  File "/home/pi/opensourceleg/opensourceleg/osl.py", line 84, in __exit__
    self._knee.stop()
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 485, in stop
    self.set_mode(mode=self.control_modes.voltage)
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 529, in set_mode
    self._mode.transition(to_state=mode)
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 180, in transition
    self.exit()
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 170, in exit
    self._exit_callback()
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 213, in _exit
    self._set_voltage(voltage=0)
  File "/home/pi/opensourceleg/opensourceleg/hardware/actuators.py", line 217, in _set_voltage
    self._device.send_motor_command(
  File "/usr/local/lib/python3.9/dist-packages/flexsea/device.py", line 424, in send_motor_command
    raise ValueError(f"Invalid device ID: {self.dev_id}")
ValueError: Invalid device ID: 1379
In 29754 cycles at 200.00 Hz:
        avg error: 0.024 milliseconds
        stddev error: 0.252 milliseconds
        percent of time sleeping: 87.4 %
pi@neurobionics:~ $ 
```

tkevinbest commented:
Here's the test script I used: 
```
"""
test_thermal_cutoffs.py
Test script to ensure that the thermal limiting functionality works on the OSL. 

Kevin Best, 6/26/2024
"""
from opensourceleg.opensourceleg.osl import OpenSourceLeg

LOOP_FREQUENCY = 200 # Hz
osl = OpenSourceLeg(frequency=LOOP_FREQUENCY, file_name='./thermal_cutoff_test')
osl.clock.report = True

osl.add_joint("knee", gear_ratio=9 * 83 / 18, offline_mode=False)

joint_vars_to_log = ["output_position", "output_velocity", "motor_position", "motor_velocity",
                          "winding_temperature", "case_temperature", "motor_current", 
                          "motor_voltage","battery_voltage","battery_current",
                          "joint_position", "joint_velocity", "thermal_scaling_factor", "joint_torque"]
osl.log.add_attributes(osl.knee, joint_vars_to_log)

osl.knee.set_max_case_temperature(30)
osl.knee.set_max_winding_temperature(35)

with osl:
    osl.home()
    osl.knee.set_mode(osl.knee.control_modes.current)
    for t in osl.clock:
        osl.update()
        osl.knee.set_output_torque(-15.0)
        print("Case {:2f}, Winding {:2f}".format(osl.knee.case_temperature, osl.knee.winding_temperature))

```

imsenthur commented:
Good catch! I went ahead and modified a few things:

1) Created a custom exception under the `safety` module to raise whenever the temperature goes past the thermal limits
2) Moved the exiting routine back to the OSL class. I believe this is a good design principle to enforce throughout the library: **components only raise exceptions and systems handle those exceptions gracefully and exit the script**. This would prevent any further multiple-exit()-statements-being-called scenario.
3) The actuators module now raises exceptions instead of acting (stopping the motor/closing the device) when errors occur (thermal limit reached), this means that it is left to the user to handle this exception if used as a standalone class. This behavior will be made better when we complete the generalization.

tkevinbest commented:
Seems legit - tested with both the case and the winding temps and system shuts down as expected.

tkevinbest commented on opensourceleg/tools/safety.py:
```diff
@@ -6,6 +6,12 @@
 import numpy as np
 
 
+class ThermalLimitExceeded(Exception):
```
Should we name this class "exception"? Not sure what is normal?
